cpp/src/io/parquet/delta_binary.cuh

```diff
@@ -179,6 +173,28 @@ struct delta_binary_decoder {
     }
   }
 
+  // given start/end pointers in the data, find the end of the binary encoded block. when done,
+  // `this` will be initialized with the correct start and end positions. returns the end, which is
+  // start of data/next block. should only be called from thread 0.
+  inline __device__ uint8_t const* find_end_of_block(uint8_t const* start, uint8_t const* end)
```

Comment:
Out of curiosity, how hard would parallelizing this be, even at the warp level?  Is this a similar thing to the level data where you can't easily do it because you can't know where position N+1 is without processing position N? It seems to do a decent amount of work and it shows up in a number of places.  I don't have a good sense how many mini blocks you might see in a typical page.

Reply:
Sadly this is just like trying to count PLAIN encoded string sizes, or as you point out trying to find where runs start in the RLE decoder.  The header for each block has a varint in it, and then each of the miniblocks can have its own encoding bitwidth. So you have to traverse the entire thing to find the end of current section/start of the next. Other implementations usually just allocate an array to hold all the values, since they're only dealing with one page per column chunk at a time.

Pages these days are usually capped at 20,000 rows (at least cudf and parquet-mr), with 128 values per block, so on the order of 157 block headers to parse and skip per page.  Of course this can get pretty big if lists are involved.